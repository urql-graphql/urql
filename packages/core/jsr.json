{
  "name": "@urql/core",
  "version": "5.0.6",
  "exports": {
    ".": "./src/index.ts",
    "./internal": "./src/internal/index.ts"
  },
  "exclude": [
    "node_modules",
    "cypress",
    "**/*.test.*",
    "**/*.spec.*",
    "**/*.test.*.snap",
    "**/*.spec.*.snap"
  ]
}
