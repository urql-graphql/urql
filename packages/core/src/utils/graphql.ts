import type * as GraphQLWeb from '@0no-co/graphql.web';
import type * as GraphQL from 'graphql';

type OrNever<T> = 0 extends 1 & T ? never : T;

export type GraphQLError =
  | GraphQLWeb.GraphQLError
  | OrNever<GraphQL.GraphQLError>;

export type DocumentNode =
  | GraphQLWeb.DocumentNode
  | OrNever<GraphQL.DocumentNode>;

export type DefinitionNode =
  | GraphQLWeb.DefinitionNode
  | OrNever<GraphQL.DefinitionNode>;
