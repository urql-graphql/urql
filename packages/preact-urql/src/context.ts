import { createContext } from 'preact';
import { useContext } from 'preact/hooks';
import { Client } from '@urql/core';

const OBJ = {};
export const Context: import('preact').Context<Client | object> =
  createContext(OBJ);
export const Provider: import('preact').Provider<Client | object> =
  Context.Provider;
export const Consumer: import('preact').Consumer<Client | object> =
  Context.Consumer;

Context.displayName = 'UrqlContext';

export const useClient = (): Client => {
  const client = useContext(Context);

  if (client === OBJ && process.env.NODE_ENV !== 'production') {
    const error =
      "No client has been specified using urql's Provider. please create a client and add a Provider.";

    console.error(error);
    throw new Error(error);
  }

  return client as Client;
};
