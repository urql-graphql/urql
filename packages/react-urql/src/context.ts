import { createContext, useContext } from 'react';
import { Client } from '@urql/core';

const OBJ = {};

/** `urql`'s React Context.
 *
 * @remarks
 * The React Context that `urql`â€™s {@link Client} will be provided with.
 * You may use the reexported {@link Provider} to provide a `Client` as well.
 */
export const Context: import('react').Context<Client | object> =
  createContext(OBJ);

/** Provider for `urql`'s {@link Client} to GraphQL hooks.
 *
 * @remarks
 * `Provider` accepts a {@link Client} and provides it to all GraphQL hooks,
 * and {@link useClient}.
 *
 * You should make sure to create a {@link Client} and provide it with the
 * `Provider` to parts of your component tree that use GraphQL hooks.
 *
 * @example
 * ```tsx
 * import { Provider } from 'urql';
 * // All of `@urql/core` is also re-exported by `urql`:
 * import { Client, cacheExchange, fetchExchange } from '@urql/core';
 *
 * const client = new Client({
 *   url: 'https://API',
 *   exchanges: [cacheExchange, fetchExchange],
 * });
 *
 * const App = () => (
 *   <Provider value={client}>
 *     <Component />
 *   </Provider>
 * );
 * ```
 */
export const Provider: import('react').Provider<Client | object> =
  Context.Provider;

/** React Consumer component, providing the {@link Client} provided on a parent component.
 * @remarks
 * This is an alias for {@link Context.Consumer}.
 */
export const Consumer: import('react').Consumer<Client | object> =
  Context.Consumer;

Context.displayName = 'UrqlContext';

/** Hook returning a {@link Client} from {@link Context}.
 *
 * @remarks
 * `useClient` is a convenience hook, which accesses `urql`'s {@link Context}
 * and returns the {@link Client} defined on it.
 *
 * This will be the {@link Client} you passed to a {@link Provider}
 * you wrapped your elements containing this hook with.
 *
 * @throws
 * In development, if the component you call `useClient()` in is
 * not wrapped in a {@link Provider}, an error is thrown.
 */
export const useClient = (): Client => {
  const client = useContext(Context);

  if (client === OBJ && process.env.NODE_ENV !== 'production') {
    const error =
      "No client has been specified using urql's Provider. please create a client and add a Provider.";

    console.error(error);
    throw new Error(error);
  }

  return client as Client;
};
