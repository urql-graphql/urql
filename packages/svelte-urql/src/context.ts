import { setContext, getContext } from 'svelte';
import { Client, ClientOptions } from '@urql/core';

const _contextKey = '$$_urql';

/** Retrieves a Client from Svelte's context */
export const getContextClient = (): Client => {
  const client = getContext(_contextKey);
  if (process.env.NODE_ENV !== 'production' && !client) {
    throw new Error(
      'No urql Client was found in Svelte context. Did you forget to call setContextClient?'
    );
  }

  return client as Client;
};

/** Sets a Client on Svelte's context */
export const setContextClient = (client: Client): void => {
  setContext(_contextKey, client);
};

/** Creates Client and adds it to Svelte's context */
export const initContextClient = (args: ClientOptions): Client => {
  const client = new Client(args);
  setContextClient(client);
  return client;
};
