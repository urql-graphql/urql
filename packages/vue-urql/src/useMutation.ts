/* eslint-disable react-hooks/rules-of-hooks */

import { ref, Ref } from 'vue';
import { DocumentNode } from 'graphql';
import { pipe, toPromise, take } from 'wonka';

import {
  Client,
  TypedDocumentNode,
  CombinedError,
  Operation,
  OperationContext,
  OperationResult,
  createRequest,
} from '@urql/core';

import { useClient } from './useClient';
import { unwrapPossibleProxy } from './utils';

export interface UseMutationState<T, V> {
  fetching: Ref<boolean>;
  stale: Ref<boolean>;
  data: Ref<T | undefined>;
  error: Ref<CombinedError | undefined>;
  extensions: Ref<Record<string, any> | undefined>;
  operation: Ref<Operation<T, V> | undefined>;
  executeMutation: (
    variables: V,
    context?: Partial<OperationContext>
  ) => Promise<OperationResult<T>>;
}

export type UseMutationResponse<T, V> = UseMutationState<T, V>;

export function useMutation<T = any, V = any>(
  query: TypedDocumentNode<T, V> | DocumentNode | string
): UseMutationResponse<T, V> {
  return callUseMutation(query);
}

export function callUseMutation<T = any, V = any>(
  query: TypedDocumentNode<T, V> | DocumentNode | string,
  client: Client = useClient()
): UseMutationResponse<T, V> {
  const data: Ref<T | undefined> = ref();
  const stale: Ref<boolean> = ref(false);
  const fetching: Ref<boolean> = ref(false);
  const error: Ref<CombinedError | undefined> = ref();
  const operation: Ref<Operation<T, V> | undefined> = ref();
  const extensions: Ref<Record<string, any> | undefined> = ref();

  return {
    data,
    stale,
    fetching,
    error,
    operation,
    extensions,
    executeMutation(
      variables: V,
      context?: Partial<OperationContext>
    ): Promise<OperationResult<T, V>> {
      fetching.value = true;

      return pipe(
        client.executeMutation<T, V>(
          createRequest<T, V>(query, unwrapPossibleProxy<V>(variables)),
          context || {}
        ),
        take(1),
        toPromise
      ).then((res: OperationResult) => {
        data.value = res.data;
        stale.value = !!res.stale;
        fetching.value = false;
        error.value = res.error;
        operation.value = res.operation;
        extensions.value = res.extensions;
        return res;
      });
    },
  };
}
