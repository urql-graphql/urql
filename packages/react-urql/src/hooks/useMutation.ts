import { DocumentNode } from 'graphql';
import { useState, useCallback, useRef, useEffect } from 'react';
import { pipe, toPromise } from 'wonka';

import {
  TypedDocumentNode,
  OperationResult,
  OperationContext,
  CombinedError,
  createRequest,
  Operation,
} from '@urql/core';

import { useClient } from '../context';
import { initialState } from './state';

export interface UseMutationState<Data = any, Variables = object> {
  fetching: boolean;
  stale: boolean;
  data?: Data;
  error?: CombinedError;
  extensions?: Record<string, any>;
  operation?: Operation<Data, Variables>;
}

export type UseMutationResponse<Data = any, Variables = object> = [
  UseMutationState<Data, Variables>,
  (
    variables?: Variables,
    context?: Partial<OperationContext>
  ) => Promise<OperationResult<Data, Variables>>
];

export function useMutation<Data = any, Variables = object>(
  query: DocumentNode | TypedDocumentNode<Data, Variables> | string
): UseMutationResponse<Data, Variables> {
  const isMounted = useRef(true);
  const client = useClient();

  const [state, setState] = useState<UseMutationState<Data, Variables>>(
    initialState
  );

  const executeMutation = useCallback(
    (variables?: Variables, context?: Partial<OperationContext>) => {
      setState({ ...initialState, fetching: true });

      return pipe(
        client.executeMutation<Data, Variables>(
          createRequest<Data, Variables>(query, variables),
          context || {}
        ),
        toPromise
      ).then(result => {
        if (isMounted.current) {
          setState({
            fetching: false,
            stale: !!result.stale,
            data: result.data,
            error: result.error,
            extensions: result.extensions,
            operation: result.operation,
          });
        }
        return result;
      });
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [client, query, setState]
  );

  useEffect(() => {
    isMounted.current = true;
    return () => {
      isMounted.current = false;
    };
  }, []);

  return [state, executeMutation];
}
