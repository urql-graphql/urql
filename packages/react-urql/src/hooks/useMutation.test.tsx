/* eslint-disable react-hooks/rules-of-hooks */
import { vi, expect, it, beforeEach, describe, beforeAll, Mock } from 'vitest';

// Note: Testing for hooks is not yet supported in Enzyme - https://github.com/airbnb/enzyme/issues/2011
vi.mock('../context', async () => {
  const { delay, fromValue, pipe } = await vi.importActual('wonka');
  const mock = {
    executeMutation: vi.fn(() =>
      pipe(fromValue({ data: 1, error: 2, extensions: { i: 1 } }), delay(200))
    ),
  };

  return {
    useClient: () => mock,
  };
});

import { print } from 'graphql';
import { gql } from '@urql/core';
import React from 'react';
import renderer, { act } from 'react-test-renderer';

import { useClient } from '../context';
import { useMutation } from './useMutation';

// @ts-ignore
const client = useClient() as { executeMutation: Mock };

const props = {
  query: 'mutation Example { example }',
};

let state: any;
let execute: any;

const MutationUser = ({ query }: { query: any }) => {
  const [s, e] = useMutation(query);
  state = s;
  execute = e;
  return <p>{s.data}</p>;
};

beforeAll(() => {
  // TODO: Fix use of act()
  vi.spyOn(global.console, 'error').mockImplementation(() => {
    // do nothing
  });
});

beforeEach(() => {
  client.executeMutation.mockClear();
  state = undefined;
  execute = undefined;
});

describe('on initial useEffect', () => {
  it('initialises default state', () => {
    renderer.create(<MutationUser {...props} />);
    expect(state).toMatchSnapshot();
  });

  it('does not execute subscription', () => {
    renderer.create(<MutationUser {...props} />);
    expect(client.executeMutation).toBeCalledTimes(0);
  });
});

describe('on execute', () => {
  const vars = { test: 1234 };

  it('sets fetching to true', () => {
    renderer.create(<MutationUser {...props} />);
    act(() => {
      execute(vars);
    });
    expect(state).toHaveProperty('fetching', true);
  });

  it('calls executeMutation', () => {
    renderer.create(<MutationUser {...props} />);
    act(() => {
      execute(vars);
    });
    expect(client.executeMutation).toBeCalledTimes(1);
  });

  it('calls executeMutation with query', () => {
    renderer.create(<MutationUser {...props} />);
    act(() => {
      execute(vars);
    });

    const call = client.executeMutation.mock.calls[0][0];
    expect(print(call.query)).toBe(print(gql(props.query)));
  });

  it('calls executeMutation with variables', () => {
    renderer.create(<MutationUser {...props} />);
    act(() => {
      execute(vars);
    });
    expect(client.executeMutation.mock.calls[0][0]).toHaveProperty(
      'variables',
      vars
    );
  });

  it('can adjust context in executeMutation', () => {
    renderer.create(<MutationUser {...props} />);
    act(() => {
      execute(vars, { url: 'test' });
    });
    expect(client.executeMutation.mock.calls[0][1].url).toBe('test');
  });
});

describe('on subscription update', () => {
  it('forwards data response', async () => {
    const wrapper = renderer.create(<MutationUser {...props} />);
    await execute();
    wrapper.update(<MutationUser {...props} />);

    expect(state).toHaveProperty('data', 1);
  });

  it('forwards error response', async () => {
    const wrapper = renderer.create(<MutationUser {...props} />);
    await execute();
    wrapper.update(<MutationUser {...props} />);

    expect(state).toHaveProperty('error', 2);
  });

  it('forwards extensions response', async () => {
    const wrapper = renderer.create(<MutationUser {...props} />);
    await execute();
    wrapper.update(<MutationUser {...props} />);

    expect(state).toHaveProperty('extensions', { i: 1 });
  });

  it('sets fetching to false', async () => {
    const wrapper = renderer.create(<MutationUser {...props} />);
    wrapper.update(<MutationUser {...props} />);

    await execute();
    wrapper.update(<MutationUser {...props} />);
    expect(state).toHaveProperty('fetching', false);
  });
});
