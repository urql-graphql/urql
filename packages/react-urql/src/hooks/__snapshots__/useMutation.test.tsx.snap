// Vitest Snapshot v1

exports[`on initial useEffect > initialises default state 1`] = `
{
  "data": undefined,
  "error": undefined,
  "extensions": undefined,
  "fetching": false,
  "operation": undefined,
  "stale": false,
}
`;
